main/src/ca/uwaterloo/flix/language/phase/optimizer/LambdaDrop.scala

```diff
@@ -0,0 +1,463 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.optimizer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.MonoAst.{Expr, empty}
+import ca.uwaterloo.flix.language.ast.shared.{BoundBy, Scope}
+import ca.uwaterloo.flix.language.ast.{MonoAst, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.dbg.AstPrinter.DebugMonoAst
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import scala.collection.mutable
+
+/**
+  * Rewrites recursive higher-order functions to non-recursive functions with
+  * a recursive local def where constant parameters become closure captured.
+  *
+  * A constant parameter is one that has the same value in every recursive call.
+  *
+  * Example:
+  *
+  * {{{
+  *   def map(f: a -> b, l: List[a]): List[b] = match l {
+  *       case Nil     => Nil
+  *       case x :: xs => f(x) :: map(f, xs)
+  *   }
+  * }}}
+  *
+  * Is rewritten to:
+  *
+  * {{{
+  *   def map(f: a -> b, l: List[a]): List[b] =
+  *       def map$loop(l1) = match l1 {
+  *           case Nil      => Nil
+  *           case x :: xs  => f(x) :: map$loop(xs)
+  *       };
+  *       map$loop(l)
+  * }}}
+  */
+object LambdaDrop {
+
+  /** See [[LambdaDrop]] for documentation. */
+  def run(root: MonoAst.Root)(implicit flix: Flix): MonoAst.Root = flix.phase("LambdaDrop") {
+    val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    root.copy(defs = newDefs)
+  }
+
+  private def visitDef(defn: MonoAst.Def)(implicit flix: Flix): MonoAst.Def = {
+    implicit val lctx: LocalContext = LocalContext.mk()
+    if (isDroppable(defn))
+      lambdaDrop(defn)
+    else
+      defn
+  }
+
+  /**
+    * A function is droppable if
+    * (a) it contains at least one recursive call and
+    * (b) is a higher-order function, i.e., it has at least one formal parameter
+    * with a function type.
+    *
+    * The latter condition can be checked by the [[isHigherOrder]] predicate.
+    */
+  private def isDroppable(defn: MonoAst.Def)(implicit lctx: LocalContext): Boolean = {
+    if (isHigherOrder(defn)) {
+      visitExp(defn.exp)(defn.sym, lctx)
+      lctx.recursiveCalls.nonEmpty
+    } else {
+      false
+    }
+  }
+
+  /** Returns `true` if at least one formal parameter of `defn` has an arrow type. */
+  private def isHigherOrder(defn: MonoAst.Def): Boolean = {
+    defn.spec.fparams.exists {
+      fp =>
+        fp.tpe.typeConstructor match {
+          case Some(TypeConstructor.Arrow(_)) => true
+          case Some(TypeConstructor.ArrowWithoutEffect(_)) => true
+          case Some(_) => true
+          case None => false
+        }
+    }
+  }
```

Comment:
@magnus-madsen 
The problem here is that I'm only checking for `Arrow` types. How do we ensure that it is a function type and not a type with a type param like `List`?

Reply:
Ah wait no that's `Type.Apply`, right? So what could be the issue here?

Reply:
Okay wait there is a `true` in the wildcard case... what a typo